### ISSUE TITLE
Can an iPython notebook be cited under USGS Fundamental Science Practices?

### ISSUE THREAD

**Issue by hjenter** [usgs/best-practices#25]
Is there a way to make an iPython notebook into a reference that can be cited officially under USGS FSP rules?

**Comment by rviger-usgs** [comment]
in IPDS, record this as a software or data release. You'll have review and metadata requirements, but it should be an official USGS information product when done. Not sure about hosting the thing. Can keep it in USGS machines (bitbucket or sciencebase) for the source, but point to some additional live rendering if one wanted to.

**Comment by rsignell-usgs** [comment]
In the new USGS Software Release Policy
https://www2.usgs.gov/usgs-manual/im/IM-OSQI-2016-01.html
it says:
 

> G. Code Distribution.   Software releases may be distributed via a third party, but the authoritative version of the approved software release and comparable information must be made publicly available through a U.S. Government platform (OMB Memorandum 10-23, Section 3).

So I guess that means that while it would nice to simply release on GitHub, where you could get a DOI easily via Zenodo (https://guides.github.com/activities/citable-code/) and where your notebook would be nicely rendered, you need to have it on a site like USGS bitbucket instead.

Is that right folks?

**Comment by rviger-usgs** [comment]
I think the DOI would point to the USGS version of the resource, but that the landing page there could point to an external site that might be more visible or provide nicer supporting services. This way, if the external site goes away for whatever reason, the USGS artifact will persist (and the landing page would get updated). Probably also good to make clear the preferred citation on the external cite (that is, point to the DOI and the USGS at the bottom of it all).

**Comment by cbdawson-usgs** [comment]
@rsignell-usgs @rviger-usgs Yes, my understanding is that the DOI would point to the .gov page, whether that be on bitbucket.usgs.gov or elsewhere on usgs.gov. 

I've gotten the DOIs for maybe a half dozen software packages so far, it's very quick and [easy to do](https://www1.usgs.gov/csas/doi/) and parallels the Data Release DOI workflow:

1. You _reserve_ the DOI when you are working on your documentation or web site, so that you can include it in the documentation. 
2. You then _register_ the DOI only when the .gov URL is live/updated. 

Because the code citation, separate and distinct from any associated report or article, is relatively new for USGS, we've also started including an explanation of what the two citations are. See our [MT3D-USGS page](http://water.usgs.gov/ogw/mt3d-usgs/#cite), for example. This is our second iteration of the language, which we're hoping to streamline.

Your .gov page can link to your github site. This is how we've dealt with the [FloPy page](http://water.usgs.gov/ogw/flopy/), which links out to flopy github repos, which include notebooks.

**Comment by mguy-usgs** [comment]
Looks good, like the helpful example

**Comment by skybristol** [comment]
I just posed some ideas about code discovery under #28 that relates here. I'm a little on the fence about where USGS DOIs should dereference - third party repo like GitHub, USGS repo like Bitbucket, some other landing page, etc. Part of the reason that we do want to use the existing DOI prefix (10.5066, or maybe a smaller subset of DOI prefixes that are actually "owned" by USGS) is so that we can more easily do citation tracking and build metrics. We do need to make sure the process of getting those DOIs minted and managed is at least almost as easy as Zenodo. If you use ScienceBase to mint a DOI, then it is that easy (or at least it could be if the permissions were delegated to use that function beyond ScienceBase admins), but that doesn't necessarily need to be the only answer.

To some extent, it makes perfect sense to have the DOI dereference right to a code repo where a user can get the code being cited. However, there are a number of considerations that I've been thinking about:

- At least in our current state of maturity, there will be a tremendous amount of variability across the "landing pages" for USGS code. That may or may not be a bad thing. In some cases, we may not care what a landing page looks like to a human, but it's probably something to think about.
- If you do have both a GitHub (or some other third party) repo where you are doing your active development and a continuous or periodic clone on a USGS host, you might want to let interested users know that both of those exist. You might want to send people to your GitHub repo if they want the latest and greatest code but also your USGS repo for the official cited version/packaging.
- There's quite a lot of information that can be retrieved programmatically via API (see my thoughts on #28) from a GitHub repo and some amount of information that could come from Atlassian's Bitbucket on myUSGS. However, there's quite a bit of variability across the various source control or software distribution platforms we use now and will expand to in the future.

For some of these reasons, we've been experimenting with creating ScienceBase Items as our core metadata store and landing page as the dereferencing point for software DOIs. Those pages can be read programmatically as ISO metadata or their native JSON to augment the DataCite metadata from the DOI system and build out various kinds of catalogs/views. They can be set up to provide a standardized landing page content and look. In general, the point with these items is to provide links to source control as opposed to housing actual code bytes, but ScienceBase could perhaps also be used as the snapshot location for code that you want to freeze and archive.

On the other hand, we could also accomplish many of these same things directly from a code repo with a little bit of thought. That's why I'm kind of on the fence.

**Comment by mguy-usgs** [comment]
An advantage of ScienceBase as @skybristol described is that it has a broader USGS audience likely than git or any other software repo and is a standard and active tool for meeting the USGS data management policy. It can be nice (but probably less powerful) to have a common tool to use/learn/search for data and software, as least at the kick off point as described.

**Comment by emartinez-usgs** [comment]
While ScienceBase is a resource within the USGS, that is about the end of its reach. If you try Google searching for USGS software, ScienceBase does not come back. You will find various fragments of the USGS and somewhere in this list you do find GitHub (about position 4 when I searched).

My point here is we should try to make our software (and the underlying science) as available as possible. To achieve this we need to consider where the broader community goes (de-facto) to look for these resources. For source code, the current answer is GitHub. In the past it was Source Forge. In the future it may be something else; we should embrace this.

With this in mind, I feel DOIs should resolve to a static resource. I think MyUSGS (BitBucket) is a solution we can use. We can set up automatic (or on-demand) syncs from the public de-facto hosting solution (currently; GitHub) back to the MyUSGS instance in order to ensure things stay up-to-date.

Landing pages on projects could (should?) be standardized within the USGS. There are some key bits of information that need be conveyed on these landing pages (project name, description, DOI, cross-links, license, disclaimer, status, etc... ?). This is not to limit the documentation or overall use of ".md" files on GitHub per-project, but rather to establish a standard and consistent beginning to the user experience.

As @skybristol suggests (and I allude to above), cross-listing the source as "mirrored" on MyUSGS and GitHub is a great idea.

**Comment by skybristol** [comment]
Well...part of the reason that you don't see ScienceBase search results in public search for "USGS software" is that there's not a whole lot of software cataloged in that resource yet. You will find a whole lot of other USGS science and scientific products coming up as a result from ScienceBase, though, and they've done a good job on SEO with schema.org and other methods that mean ScienceBase stuff comes up in the wild quite a lot. We don't have quite the same luxury to do SEO things with Atlassian Bitbucket via myUSGS, and I would not particularly recommend that being the authoritative, usgs.gov-based landing locale that we try to make work for us. I don't really care if ScienceBase is in the mix or not, though it might be a reasonable engine to use as part of a solution.

As I've said elsewhere, I made a run at getting code.usgs.gov as a domain for an app that we could build out in a consistent way, starting from the DOI registration requirement we put in the policy. I wasn't successful in that given the current posture of the WRET, but I may come back from a holiday break with a renewed zeal for tilting at that windmill. I think it would be pretty sweet if we had an app there with consistent landing pages at something like /code.usgs.gov/doi/10.5066/12345 that respond to content negotiation for HTML, JSON, whatever outputs. We can do SEO stuff on the HTML pages and the web portal to improve public search and do linked data stuff (triple store and SPARQL end point as one access mechanism) with the persistent identifiers to promote other kinds of linked data apps against the resource. We can go back and mint DOIs for code projects and products that pre-date the new policy to get those in the mix.

The reality is that not everyone is going to use GitHub for managing their projects. Not everyone is going to use myUSGS Bitbucket (or whatever we maintain there over time) as their .gov alternative access point. I think we need a layer of abstraction with a catalog/index of some kind on top of the heterogenous source control and project management spaces for USGS code projects. We've required the DOI, so that's the one common point of reference in the whole system. We can work with DataCite metadata dynamically via API, so we don't have to ask anyone to do some other step. And wherever possible, when those DOIs eventually get us to something like a GitHub repo, we can write code that augments the basic DataCite metadata with a lot more info to improve discoverability.
